quizx/src/flow/causal.rs

```diff
@@ -132,10 +133,41 @@ impl CausalFlow {
         self.lines[p.line].get(p.pos + 1).copied()
     }
 
+    /// Returns the previous node in the causal flow.
+    pub fn pred(&self, v: V) -> Option<V> {
+        let p = self.positions[v];
+        p.pos.checked_sub(1).map(|pos| self.lines[p.line][pos])
+    }
+
+    pub fn inputs(&self) -> impl Iterator<Item = V> + '_ {
+        self.lines.iter().map(|line| line[0])
+    }
+
+    pub fn outputs(&self) -> impl Iterator<Item = V> + '_ {
+        self.lines.iter().map(|line| *line.last().unwrap())
+    }
+
+    /// The line index of a vertex.
+    pub fn line_idx(&self, v: V) -> usize {
+        self.positions[v].line
+    }
+
     /// Returns the causal flow lines of the graph.
     pub fn lines(&self) -> &[Vec<V>] {
         &self.lines
     }
+
+    /// Whether u -> v is a causal edge.
+    pub fn is_causal_edge(&self, u: V, v: V) -> bool {
+        self.next(u) == Some(v)
+    }
+
+    /// Whether (u, v) is a non-causal edge.
+    ///
+    /// This is true when both u -> v and v -> u are not causal.
+    pub fn is_non_causal_edge(&self, u: V, v: V) -> bool {
```

Comment:
A bit weird that `¬ (P ∧ ¬P ⊸ ⊤)` :P

Reply:
true, feel free to suggest another name for "non-causal" edges :)

Reply:
how about I rename `is_causal` to `is_causal_dir` and then `is_causal` is true if the edge is causal in either direction